svelte/preprocess/src/index.ts

```diff
@@ -0,0 +1,96 @@
+import MagicString from 'magic-string';
+import type {PreprocessorGroup} from 'svelte/compiler';
+import {parse} from 'svelte/compiler';
+import type {TemplateNode, Attribute} from 'svelte/types/compiler/interfaces';
+
+export const directivesPreprocess = (): PreprocessorGroup => {
+	return {
+		name: 'AgnosUI',
+		markup: async ({content, filename}) => {
+			const varPrefix = '__AgnosUISveltePreprocess__';
+			const str = new MagicString(content, {filename});
+			const parsedCode = parse(content, {filename});
+			const requiredImports = new Set<string>();
+
+			const extractValue = (attribute: Attribute) => {
+				const res: string[] = [];
+				const value = attribute.value;
+				for (const part of value) {
+					if (part.type === 'Text') {
+						res.push(JSON.stringify(part.data));
+					} else if (part.type === 'MustacheTag') {
+						res.push(`(${content.substring(part.expression.start, part.expression.end)})`);
+					} else {
+						throw new Error(`Unexpected part type: ${part.type}`);
+					}
+				}
+				return res.join('+');
+			};
+
+			const processItem = (item: TemplateNode) => {
+				const actionAttributes = [];
+				const classAttributes = [];
+				if (item.attributes) {
+					for (const attribute of item.attributes) {
+						if (attribute.type === 'Action') {
+							actionAttributes.push(attribute);
+						} else if (attribute.type === 'Attribute' && attribute.name === 'class') {
+							classAttributes.push(attribute);
+						}
+					}
+					if (actionAttributes.length > 0) {
+						const end = item.attributes[item.attributes.length - 1].end;
+						let first = true;
+						str.appendRight(end, ` {...(${varPrefix}BROWSER ? {} : ${varPrefix}ssrAttributes(`);
+						requiredImports.add('ssrAttributes');
+						for (const attribute of actionAttributes) {
+							if (first) {
+								first = false;
+							} else {
+								str.appendRight(end, `, `);
+							}
+							if (attribute.expression) {
+								str.appendRight(end, `[${attribute.name}, ${content.substring(attribute.expression.start, attribute.expression.end)}]`);
+							} else {
+								str.appendRight(end, attribute.name);
+							}
+						}
+
+						for (const attribute of classAttributes) {
+							const value = extractValue(attribute);
+							str.overwrite(attribute.start, attribute.end, `use:${varPrefix}classDirective={${value}}`);
+							str.appendRight(end, `, [${varPrefix}classDirective, ${extractValue(attribute)}]`);
+							requiredImports.add('classDirective');
+						}
+
+						str.appendRight(end, `))}`);
+					}
+				}
+				item.children?.forEach(processItem);
```

Comment:
wouldn't maybe better to use [walk](https://svelte.dev/docs/svelte-compiler#walk) for performances?

Reply:
`walk` does not seem to be compatible with the html part of svelte files, only the script parts.